contracts/Strategy.sol

```diff
@@ -145,67 +208,43 @@ contract Strategy is BaseStrategy {
         }
 
         // put want into lp then put want-lp into masterchef
-        uint256 pooledBefore = balanceOfPooled();
-        uint256 amountIn = Math.min(maxSingleDeposit, balanceOfWant());
-        if (joinPool(amountIn, assets, numTokens, tokenIndex, balancerPoolId)) {
+        if (_joinPool()) {
             // put all want-lp into masterchef
             masterChef.deposit(masterChefPoolId, balanceOfBpt(), address(this));
-
-            uint256 pooledDelta = balanceOfPooled().sub(pooledBefore);
-            uint256 joinSlipped = amountIn > pooledDelta ? amountIn.sub(pooledDelta) : 0;
-            uint256 maxLoss = amountIn.mul(maxSlippageIn).div(basisOne);
-            require(joinSlipped <= maxLoss, "Slipped in!");
-            lastDepositTime = now;
         }
-
-        // claim all beets
-        claimRewards();
-        // and stake all
-        stakeAllRewards();
     }
 
     function liquidatePosition(uint256 _amountNeeded) internal override returns (uint256 _liquidatedAmount, uint256 _loss){
-        if (estimatedTotalAssets() < _amountNeeded) {
-            _liquidatedAmount = liquidateAllPositions();
-            return (_liquidatedAmount, _amountNeeded.sub(_liquidatedAmount));
-        }
-
         uint256 looseAmount = balanceOfWant();
         if (_amountNeeded > looseAmount) {
-            uint256 toExitAmount = _amountNeeded.sub(looseAmount);
+            // convert amount still needed to bpt
+            uint256 toExitAmount = tokensToBpts(_amountNeeded.sub(looseAmount));
 
-            // withdraw all bpt out of masterchef
-            masterChef.withdrawAndHarvest(masterChefPoolId, balanceOfBptInMasterChef(), address(this));
-            // sell some bpt
-            exitPoolExactToken(toExitAmount);
-            // put remaining bpt back into masterchef
-            masterChef.deposit(masterChefPoolId, balanceOfBpt(), address(this));
+            // withdraw needed bpt out of masterchef and sell it for want
+            _withdrawFromMasterChefAndSellBpt(toExitAmount);
 
             _liquidatedAmount = Math.min(balanceOfWant(), _amountNeeded);
             _loss = _amountNeeded.sub(_liquidatedAmount);
-
-            _enforceSlippageOut(toExitAmount, _liquidatedAmount.sub(looseAmount));
         } else {
             _liquidatedAmount = _amountNeeded;
         }
     }
```

Comment:
Seems like there should be a deposit any extra loose BPTs back into the masterchef

Reply:
This will happen in the adjustPosition if we can join the pool. Isn't that better than doing it everytime? Maybe we just wanna leave the pool and we keep staking.